apis/apiextensions/fn/v1alpha1/functionio_types.go

```diff
@@ -0,0 +1,266 @@
+	Desired Desired `json:"desired,omitempty"`
+
+	// Items is a list of Crossplane resources - either XRs or MRs.
+	//
+	// A function will read this field in the input FunctionIO and populate
+	// this field in the output FunctionIO.
+	// +kubebuilder:validation:EmbeddedResource
+	// +kubebuilder:pruning:PreserveUnknownFields
+	Items []runtime.RawExtension `json:"items"`
+
+	// Results is an optional list that can be used by function to emit results
+	// for observability and debugging purposes.
+	// +optional
+	Results []Result `json:"results,omitempty"`
+}
+
+// Observed state at the beginning of a function pipeline invocation.
+type Observed struct {
+	// Composite reflects the observed state of the XR this function reconciles.
+	Composite ObservedComposite `json:"composite"`
+
+	// Resources reflect the observed state of any extant composed resources
+	// this function reconciles. Only composed resources that currently exist in
+	// the API server (i.e. have been created and not yet deleted) are included.
+	// +optional
+	Resources []ObservedResource `json:"resources,omitempty"`
+}
+
+// An ObservedComposite resource.
+type ObservedComposite struct {
+	// Resource reflects the observed XR.
+	Resource runtime.RawExtension `json:"resource"`
+
+	// ConnectionDetails reflects the observed connection details of the XR.
+	ConnectionDetails []ExplicitConnectionDetail `json:"connectionDetails"`
+}
+
+// An ObservedResource represents an observed composed resource.
+type ObservedResource struct {
+	// Name of the observed resource. Must be unique within the array of
+	// observed resources. Corresponds to the name entry in a Composition's
+	// resources array, and the name entry in the desired resources array.
+	Name string `json:"name"`
+
+	// Resource reflects the observed composed resource.
+	Resource runtime.RawExtension `json:"resource"`
+
+	// ConnectionDetails reflects the observed connection details of the
+	// composed resource.
+	ConnectionDetails []ExplicitConnectionDetail `json:"connectionDetails"`
+}
+
+// Desired state of a function pipeline invocation.
+type Desired struct {
+	// Composite reflects the desired state of the XR this function reconciles.
+	// +optional
+	Composite DesiredComposite `json:"composite,omitempty"`
+
+	// Resources reflect the desired state of composed resources, including
+	// those that do not yet exist.
+	// +optional
+	Resources []DesiredResource `json:"resources,omitempty"`
+}
+
+// A DesiredComposite resource.
+type DesiredComposite struct {
+	// TODO(negz): Presumably functions can't mutate _any_ XR field, e.g. its
+	// deletion timestamp, name, etc?
+
+	// Resource reflects the desired XR. Functions may update the metadata,
+	// spec, and status of an XR.
+	// +optional
+	Resource *runtime.RawExtension `json:"resource,omitempty"`
+
+	// ConnectionDetails reflects the desired connection details of the XR.
+	// +optional
+	ConnectionDetails []ExplicitConnectionDetail `json:"connectionDetails"`
```

Comment:
I'd prefer not to. Similar rationale to https://github.com/crossplane/crossplane/pull/3465#discussion_r1070456314, but I think it's good for the API here to be reasonably symmetrical with `DerivedConnectionDetail` (it would be strange to me for one variant to be a map and the other to be a list of named subobjects).